### PR TITLE
Add incremental security-fix history mining skill

### LINKED ISSUE #726: Mine git history for security fixes that never got an advisory
`advisories` fetches published GHSA/CVE for a repo and `advisory-deep-dive` re-audits each fix commit for regressions and bypasses. Both depend on an advisory having been published. Many security fixes never get one; the commit message says "fix bounds check" or nothing at all.

Add a `history` skill that mines `git log` directly:

- keyword-filter commit messages and titles for security-shaped language, tuned per language/ecosystem
- for candidates, read the diff (batched, 3-5 diffs per model call, size-capped) and classify: security fix / not / unclear
- for confirmed fixes, record `{commit, title, description, code_paths, vuln_type, cve_if_any}` in the same shape `advisory-deep-dive` already reads
- cache by analysed HEAD; on re-run, verify the cached HEAD is still an ancestor (force-push/rebase invalidates) and only process new commits
- shallow clones proceed but mark the output partial so later stages don't read "no history for this file" as "clean"

Output feeds `advisory-deep-dive` alongside the ecosyste.ms advisories, and the `threat-model` contract's historical-vuln section.

Source: [google/mantis](https://github.com/google/mantis) `mantis-history/SKILL.md` (Apache-2.0). Their note: "many security fixes do not get CVEs, and many security vulnerabilities are fixed without realizing they were vulnerabilities."

Related: #45 (CPAN advisory coverage; this partly compensates for ecosystems with thin advisory data).

### PR DESCRIPTION
**Pull request by abhinavgautam01**
Fixes #726

## Summary

Adds a bundled `history` skill that mines repository Git history for security fixes that may never have received a public advisory.

The skill uses deterministic, ecosystem-aware commit filtering followed by model review of bounded diff batches. Its cumulative report feeds historical vulnerability context into `threat-model` and gives `advisory-deep-dive` additional patch anchors.

## Changes

- Add the schema-backed `history` skill.
- Add ecosystem-specific security commit-message filters for C/C++, Rust, Go, Python, JavaScript/TypeScript, JVM, Ruby, PHP and .NET repositories.
- Inspect candidate commits in batches of three to five with per-diff size limits.
- Record confirmed fixes with commit, title, description, affected code paths, vulnerability type and optional CVE/GHSA identifier.
- Reuse previous reports only when the cached analyzed HEAD remains an ancestor of the current HEAD.
- Invalidate cached results after rewritten or force-pushed history.
- Mark shallow clones, truncated candidate sets and incomplete classifications as partial.
- Add `history` to code-repository triage and grouped diff rescans.
- Feed mined fixes into the threat model's historical vulnerability section.
- Let `advisory-deep-dive` inspect history-only fixes without creating synthetic advisory audit records or certificates.
- Add schema, helper-script, pipeline and metadata regression tests.
- Update bundled skill documentation.

## Tests

- `go test ./...`
- `go test -race ./...`
- `go test -race -tags evals ./internal/evals/...`
- `go vet ./...`
- `go vet -tags evals ./internal/evals/...`
- `go vet -tags podman ./...`
- `golangci-lint v2.12.2`
- `git diff --check`

### CODE REVIEW

File: skills/history/scripts/history_candidates.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+            candidates.append({"commit": commit, "title": subject, "matched_terms": sorted(set(matched))})
+    total = len(candidates)
+    candidates = candidates[: args.max_candidates]
+    shallow = git(args.repo, "rev-parse", "--is-shallow-repository").stdout.strip() == "true"
+    result = {
+        "head": head,
+        "requested_base": args.base or None,
+        "cache_reusable": reusable,
+        "cache_invalid_reason": invalid_reason,
+        "range": revision,
```
[the remark this answers is not in the document]
Comment (abhinavgautam01) [reply skills/history/scripts/history_candidates.py:193]: `cache_invalid_reason` now emits an empty JSON string when the cache is reusable, while invalid-cache cases continue to carry their explanatory reason.

I also strengthened the regression test to inspect the raw decoded JSON value and assert that this field is a string, since unmarshalling `null` into the existing Go string field did not expose the contract mismatch.

File: skills/threat-model/SKILL.md [hunk 2]
```diff
@@ -2,13 +2,14 @@
...
 license: MIT
-compatibility: Reads ./src only. Optionally pulls repo-overview and advisories from the scrutineer API for context. No registry or upstream network access required.
+compatibility: Reads ./src only. Optionally pulls repo-overview, advisories, and mined security-fix history from the scrutineer API for context. No registry or upstream network access required.
 metadata:
   scrutineer.version: 1
   scrutineer.output_file: report.json
   scrutineer.output_kind: threat_model
   scrutineer.requires:
     - recon
+    - history
```
Comment (alexandre-daubois) [inline skills/threat-model/SKILL.md]: `history` as a hard prereq means a failed history scan makes `threat-model` fail via the dead-prereq path in `preflightSkill`, and the deep-dive fan-out never happens. Same for `advisory-deep-dive`. The prose already treats a missing history report as normal, so `scrutineer.requires` should too.

**Review by alexandre-daubois — COMMENTED** [review]
Just one suggestion

**Review by andrew — CHANGES_REQUESTED** [review]
Adding `history` to `scrutineer.requires` on `threat-model` and `advisory-deep-dive` means a failed history scan fails both via the dead-prereq path in `preflightSkill` (`internal/worker/preflight.go:55-59`), and threat-model failing kills the deep-dive fan-out. Both skills already handle a missing or empty history report in prose ("carry on without it"), so drop `history` from both `scrutineer.requires` lists and update `TestBundledHistoryMetadata` / `TestBundledReconPipelineMetadata` to match. Threat-model may then dispatch before history on a fresh repo and miss the data once; a rescan picks it up, which is preferable to one history run taking the pipeline down. Keep `history` in the triage list and `enqueueDiffRescanGroup` order so it's queued ahead of threat-model in the common case.

On the changelog: please don't add an entry. `CHANGELOG.md` is written by maintainers after merge, grouping related changes and wording them for the funder audience rather than per-PR.

**Comment by abhinavgautam01** [comment]
Thanks both, fixed.

I removed `history` from the hard `scrutineer.requires` lists for both `threat-model` and `advisory-deep-dive`. Their existing required inputs remain unchanged: threat-model still requires `recon` and advisory-deep-dive still requires `advisories`.

History remains queued ahead of threat-model in both triage and grouped diff rescans, but consumers now treat its report as best-effort context as their instructions already describe. A failed history scan therefore no longer blocks threat-model, deep-dive fan-out or advisory-deep-dive.

I updated the metadata tests to pin those exact prerequisite lists.

**Review by andrew — CHANGES_REQUESTED** [review]
The candidate ceiling can permanently strand later matches. `history_candidates.py` sorts candidates oldest-first and keeps only the first 200. A complete rescan discards the partial cache and repeats those same candidates, while a reusable shallow cache at `HEAD` processes nothing. Add pagination or a continuation commit so later candidates are reviewed, with a regression test covering a history that exceeds the limit.

**Comment by abhinavgautam01** [comment]
Thanks, fixed.
`history_candidates.py` list now supports deterministic cursor pagination through --after, returning page_offset and next_cursor whenever additional candidates remain.

The skill now processes pages until the final page. If a run stops after a fully classified page, it stores that page’s continuation commit as analyzed_head instead of HEAD. The next run therefore resumes from the remaining history rather than repeating the first 200 candidates or incorrectly treating the cache as complete.

I added regression coverage using more candidates than the configured page limit. It verifies both same-run pagination without duplicates and subsequent-run continuation from the cached commit.

**Review by andrew — CHANGES_REQUESTED** [review]
Please do not reuse `next_cursor` as the next scan’s `analyzed_head`. On a merged history, the cursor commit may not contain candidates already reviewed on sibling branches. Passing that commit back as `--base` reintroduces those candidates and can alternate between branches indefinitely, leaving later candidates unreviewed. Persist the original base and `--after` cursor separately, and add a merged-history regression test.

**Comment by abhinavgautam01** [comment]
Thanks, fixed.

Pagination state now preserves the original range base and the `--after` cursor separately. `analyzed_head` pins the exact range head being processed and is no longer replaced with `next_cursor`.

The candidate helper now supports a validated `--head` argument so a later scan resumes against the same range and ordering, even if the checkout has advanced. Once that pinned range is complete, any newer commits are processed separately from `analyzed_head..HEAD`.

I also added a merged-history regression test with candidates on sibling branches. It demonstrates that using the cursor as the base reintroduces an already-reviewed sibling commit, while the new `{base, after}` continuation processes every candidate exactly once. Invalid continuation heads, bases and cursors now fail safely and trigger a fresh scan.

**Review by andrew — APPROVED** [review]
Third-round request applied correctly: `analyzed_head` stays pinned, `continuation: {base, after}` is separate, `--head` fixes the range so resume sees the same candidate ordering, and `TestHistoryCandidates_mergedHistoryKeepsBaseAndCursorSeparate` shows cursor-as-base reintroducing a reviewed sibling while the new continuation processes each candidate once.

`cache_state`'s error strings still say "cached analyzed_head" when `--base` may now be `continuation.base`; cosmetic, not blocking.
